Update to Avro 1.10.0

mcwhitak opened the pull request:
Looking for any early feedback and preparing this for the formal release of Avro `1.10.0`. Currently they [have release candidates available](https://github.com/apache/avro/releases/tag/release-1.10.0-rc2) and looking at the history of their releases this means the formal release should be in the next few weeks. I'm not suggesting we merge this before then though, I'll leave this PR open and maintain it as other changes come in until it is released.

Main changes that affect the plugin:
* Deprecation of `dateTimeLogicalType` all objects now use `JSR-310` classes
* Addition of `optionalGettersForNullableFieldsOnly` option

For `optionalGettersForNullableFieldsOnly` the option only goes into effect when `generateOptionalGetters` is `true`. As a result the testing matrix is a lot bigger than in other cases as I wanted to ensure that the behavior was performing as expected. I can scale this test case back if you think it's unnecessary.

I listed support as being at a minimum of `1.10.0` due to the option changes listed above, though one of the reasons I'm making this PR so early is in case you'd rather take a different approach.

As a follow up, when Avro does release `1.10.0` the only change necessary should be the removal of the apache staging repository.

davidmc24 commented:
Thanks for the work on this. I should have time to take a detailed look in the next day or two and get you some feedback.

davidmc24 commented:
I've done an initial review of the changes between Avro 1.9.2 and Avro 1.10.0-rc2.  Putting my notes here prior to reviewing your changes:

* As you noted, support for `dateTimeLogicalType` has been removed; with the current architecture of the plugin, that's going to be a breaking change.  Going forward, new versions of the plugin will require Avro 1.10.0+ and previous versions of the plugin will support Avro 1.9.x at most; this should be noted in the README.  My hope is that in the next major version of the plugin, I'll be using a different architecture that supports a single version of the plugin being compatible with **all** versions of Avro, but that's a ways off and may never materialize, and it would be too much effort to be worth it to support this type of version cross-compatibility as-is.
* As you noted, `SpecificCompiler` has support for a new configurable option, `optionalGettersForNullableFieldsOnly`, which we should add support for in the plugin.  This will also want test coverage in `OptionsFunctionalSpec`.
* The `Schema` class added a new overload of `toString`: `public String toString(Collection<Schema> referencedSchemas, boolean pretty)`.  It allows generating JSON schema based on Schema objects (which was already supported by other overload) but now "without inlining the referenced schemas".  This might be a capability that some users would be interested in accessing from `GenerateAvroSchemaTask`.  However, this method is deprecated, with a note that it should go away with [AVRO-2832](https://issues.apache.org/jira/browse/AVRO-2832), which is calling for a schema formatter feature.  Given that, I think we should opt not to use this method right now.
* `lang/java/compiler/src/main/velocity/org/apache/avro/compiler/specific/templates/java/classic/record.vm` was updated; since `src/test/resources/com/commercehub/gradle/plugin/avro/record.vm` in the plugin is derived from this, it may be worth updating it.  It's used in `OptionsFunctionalSpec` "supports configuring templateDirectory" and contains an added comment "// Custom template" as a marker for the test.
* It looks like two new default logical types are supported, `local-timestamp-millis` and `local-timestamp-micros`.  I don't think the plugin currently has examples/test coverage for the different default logical types (so nothing to neccesarily update as part of the Avro upgrade), but I might want to add that at some point.  There have been a few issues submitted related to people getting the syntax wrong.

davidmc24 commented:
First-pass review of your changes:

* Avro compatibility changes look good; agree with approach of leaving PR open until official Avro 1.10.0 release.  At that point we'd remove the Apache staging repository, merge, and release.
* README documentation changes look good.
* Defaults for new options should generally match the default in `SpecificCompiler`, which you did with `optionalGettersForNullableFieldsOnly`; 👍
* Changes to remove `dateTimeLogicalType` and add `optionalGettersForNullableFieldsOnly` look good.
* Now that we don't need to provide a second argument to `SpecificCompiler`'s constructor, let's update `GenerateAvroJavaTask` to remove `compile(Protocol protocol, File sourceFile)` and `compile(Schema schema, File sourceFile)`; we can just instantiate `SpecificCompiler` in the methods that would now call `compile(SpecificCompiler compiler, File sourceFile)`.
* Test support for `optionalGettersForNullableFieldsOnly`: Good start.  I recommend considering a couple minor changes.  I have no problem with how big the matrix is for the new test.  Since we can't test `optionalGettersForNullableFieldsOnly` properly without checking its interations with `gettersReturnOptional`, and we're effectively fully testing `gettersReturnOptional` as well, it may make sense to rename the test to indicate that both are tested, and remove "supports configuring gettersReturnOptional to #gettersReturnOptional".  To ensure full coverage, instead of assertions like `content.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedNullableOptionalGetter`, we may want to instead let the expectedWHATEVER flag control which expression to check for, rather than whether to check.  For example, something like `expectedNullableOptionalGetter ? content.contains("public Optional<java.lang.String> getFavoriteColor()") : content.contains("public java.lang.String getFavoriteColor()`.  That logic can probably apply to a few other tests in the class, but there's no need to update others unless you're feeling ambitious.
* It looks like you already did an update to `record.vm`; there's a bit more that could be updated to maintain parity with the latest version, but doesn't appear to break anything, so doesn't really matter.
* Please add entries to "Unreleased" section of `CHANGES.md` indicating:
  * Built using Avro 1.10.0, and crediting yourself for the contribution if desired
  * Dropped support for Avro 1.9.x
  * Removed support for `dateTimeLogicalType`
  * Added support for `optionalGettersForNullableFieldsOnly`

Overall, great work.  Honestly, this is good enough that it could be merged as-is without any changes if the new version of Avro was out already.

mcwhitak commented:
I've made the changes, some notes on their implementation
1. The test name was so long for the combined test case that I ended up having to use String continuation in order to pass the checkstyle. This doesn't look great when reading the test name in the code (it spills into the next line awkwardly due to whitespace preservation), let me know on any suggestion you have to make this look better.

2. For testing the existence of the normal getter(`public java.lang.String getFavoriteColor()`), I had initially attempted to do this (and backport the check to all other test cases using getters) however the inner builder contains the same named method, with little to distinguish it from the top level getter. My one thought was that we could use a regex to check if it was before or after the builder definition, though that felt a bit too clever (and would not have been immediately clear to someone reading the test case). Alternatively we could check how many times it appears (1 vs 2) though that feels a bit hacky as well. Let me know what you'd like to do about this.

davidmc24 commented on src/test/groovy/com/commercehub/gradle/plugin/avro/OptionsFunctionalSpec.groovy:
```diff
@@ -191,13 +186,15 @@ class OptionsFunctionalSpec extends FunctionalSpec {
     }
 
     @Unroll
```
```suggestion
    @SuppressWarnings("LineLength")
    @Unroll
```

davidmc24 commented on src/test/groovy/com/commercehub/gradle/plugin/avro/OptionsFunctionalSpec.groovy:
```diff
@@ -191,13 +186,15 @@ class OptionsFunctionalSpec extends FunctionalSpec {
 
     @Unroll
-    def "supports configuring gettersReturnOptional to #gettersReturnOptional"() {
+    def "supports configuring gettersReturnOptional to #gettersReturnOptional in conjunction with \
+setting optionalGettersForNullableFieldsOnly to #optionalGettersForNullableFieldsOnly"() {
```
```suggestion
    def "supports configuring gettersReturnOptional/optionalGettersForNullableFieldsOnly to #gettersReturnOptional/#optionalGettersForNullableFieldsOnly"() {
```

davidmc24 commented on src/test/groovy/com/commercehub/gradle/plugin/avro/OptionsFunctionalSpec.groovy:
```diff
@@ -208,17 +205,24 @@ class OptionsFunctionalSpec extends FunctionalSpec {
-        and: "the specified createOptionalGetters is used"
-        content.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedPresent
+        and: "the specified optionalGettersForNullableFieldsOnly is used"
+        content.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedNullableOptionalGetter
+        content.contains("public Optional<java.lang.String> getName()") == expectedRequiredOptionalGetter
```
```suggestion
        expectedNullableOptionalGetter ? content.contains("public Optional<java.lang.String> getFavoriteColor()") : content.contains("public java.lang.String getFavoriteColor()")
        expectedRequiredOptionalGetter ? content.contains("public Optional<java.lang.String> getName()") : content.contains("public java.lang.String getName()")
```

mcwhitak replied:
I've implemented this change but want to clarify my original comment as I'm not sure this test criteria has the intended behavior.

The inner builder that avro generates for the POJOs will always contain an unmodified `public <type> getX()` regardless of the options relating to top level getters. As a result the following condition in the above test case will pass for every configuration of both `gettersReturnOptional` and `optionalGettersForNullableFieldsOnly`

```
content.contains("public java.lang.String getFavoriteColor()")
content.contains("public java.lang.String getName()")
```

The existing tests were attempting to ensure that Optional getters were not created rather than testing for the existence of the standard getters as the standard getters always exist on a naive contains (due to the inner builder class).

So the condition of `content.contains("public java.lang.String getX()")` isn't going to provide any greater testing confidence than omitting that line entirely (as it will always be true).

davidmc24 replied:
Ooh. Good point. I’ll come up with some way to correct for that.

davidmc24 replied:
I think the way to go is along the lines of the regex suggestion you originally made.  Clear method and variable naming will make it clear enough to a casual reader.  It may break occasionally when they change their templates, but in most cases it should break in a way that causes false test failures, rather than false test passes (which are worse).  I think the regex is less likely to break than something based on line positions or match counts.  I also think that this sort of quick approach is better than pulling in a dependency on some form of source or bytecode parsing library, as the complexity isn't warranted for this use case.

What I'm thinking is along the lines of having each method that currently has assertions against `content` include a line immediately after loading `content`: `def mainClassContent = getMainClassContent(content)`.  The method would look like this:

```groovy
    /**
     * Returns just the portion of a file that relates to the main class.
     * This is used in order to allow assertions on the getters/setters/fields of the generated class itself, as opposed to a Builder.
     * 
     * @param content the file content for which to get the main content
     * @return the content of the class, from the start of the class body to the first inner class definition
     */
    @SuppressWarnings("LineLength")
    private static String getMainClassContent(String content) {
        def className = "User"
        def matcher = content =~ /(?s)public class ${className} extends org\.apache\.avro\.specific\.\SpecificRecordBase implements org\.apache\.avro\.specific\.SpecificRecord \{(?<mainClassContent>.*)public static class Builder/
        assert matcher.find()
        return matcher.group("mainClassContent")
    }
```

And then the assertions would use `mainClassContent` rather than `content`, sort of like this:

```groovy
        mainClassContent.contains("public Optional<java.lang.String> getFavoriteColor()") == expectedNullableOptionalGetter
        mainClassContent.contains("public java.lang.String getFavoriteColor()") != expectedNullableOptionalGetter
        mainClassContent.contains("public Optional<java.lang.String> getName()") == expectedRequiredOptionalGetter
        mainClassContent.contains("public java.lang.String getName()") != expectedRequiredOptionalGetter
```

mcwhitak replied:
I can make this change as part of this PR, I may not get to it for a few days but since we're waiting on the Avro release anyways seems like a good place to do it.

davidmc24 replied:
Sounds good. Please do. No hurry. As you say, we’re going to be blocked waiting for the release any way. Thanks for your work on this. You’ve been a big help.

mcwhitak replied:
I didn't realize you had written an accurate regex and properly compiling method. This was essentially a drop in, though I changed any other tests that concerned themselves specifically with conditions of the main body to use `mainClassContent` as well.

mcwhitak commented:
This should be good to go now, I've merged in master, removed apache staging and cleaned up a duplicate condition in a test case.

This release has me a bit confused though. I'm not sure if it's just the apache contributor handling this release isn't doing it with as much fanfare as normal or if something is off in their process.

1. They have the `1.10.0` [release on Github](https://github.com/apache/avro/releases/tag/release-1.10.0) but with no release notes
2. [Maven Central](https://mvnrepository.com/artifact/org.apache.avro/avro/1.10.0) has a listing for `1.10.0`
3. [JCenter](https://bintray.com/bintray/jcenter/org.apache.avro%3Aavro#) has no listing for the POM however they clearly have the files as they are [resolved properly](https://jcenter.bintray.com/org/apache/avro/avro/1.10.0/avro-1.10.0.jar) 
4. [Avro Documentation](https://avro.apache.org/docs/current/) still lists `1.9.2` as the latest

That being said, technically the dependency is resolvable from JCenter and is working as expected. I bring this up as I'd understand if you wanted to wait to merge this until everything on their side is in sync.

davidmc24 commented on src/test/groovy/com/commercehub/gradle/plugin/avro/FunctionalSpec.groovy:
```diff
@@ -72,7 +72,7 @@ abstract class FunctionalSpec extends Specification {
 
     protected void addDefaultRepository() {
-        buildFile << "repositories { jcenter() }\n"
+        buildFile << "repositories { jcenter()\n}\n"
```
Is there a reason a newline is needed here?

mcwhitak replied:
No this is a mistake, will fix

davidmc24 commented:
@mcwhitak Thanks for the updates and notification of the apparently in-progress release.  While we *could* release more or less right now, I think it may make sense to wait for a bit.  If they pulled the trigger too soon accidentally, it's possible they might try to delete and later re-publish, or almost immediately follow up with a subsequent release.  Not horribly likely, but possible.  If they haven't synced by the end of the week, I'll release Monday.
